docs(k8s): ext dns and cert manager example (#3988)
garden-io/garden: docs/advanced/cert-manager-integration.md

    kaniko:
      namespace: null
    # tlsCertificates:
    #   - name: staging-cert
    #     secretRef:
    #       name: staging-cert
    deploymentRegistry:
      hostname: "${var.registryHostname}"
      namespace: "${var.registryNamespace}"
    imagePullSecrets:
    - name: regcred
      namespace: default # Leave this in default to make configuration easier.
```

To get your context name you can simply use:

```bash
kubectl config get-contexts
kubectl config current-context
```

Now you can copy and paste your context to your `provider.context` parameter.

At this stage of the demo we are going to leave the `tlsCertificates` block commented out and we will revisit it later in a different section of this project.

For the `deploymentRegistry` and `imagePullSecrets` we require that you already have a Docker Registry (prerequisite #5). If you already have a Docker registry on e.g. Docker Hub or Qua please create a secret with your credentials using the following command:

```bash
kubectl create secret docker-registry regcred \
  --docker-username=user \
  --docker-password=password \
  --docker-email=docker-email@email.com \
  --docker-server=your-docker-server-url
```

> Note: *If you are using Dockerhub*, --docker-server should be: `https://index.docker.io/v1/`

Reviewer: ```suggestion
Note: *If you are using Docker Hub*, `--docker-server` should be: `https://index.docker.io/v1/`
```

Or, you can also follow this awesome guide from the official [Kubernetes documentation](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/). Make sure to name your secret `regcred` and keep it in the `default` namespace.

> If you decide to setup the secret with a different name make sure to update it in the `imagePullSecrets` configuration.

Reviewer: ```suggestion
If you decide to set up the secret with a different name make sure to update it in the `imagePullSecrets` configuration.
```

Setup is a noun as in, "look at my sick setup!"
"Set up" is your action if you want to set something up

> Now let's add some variables to the `project.garden.yml` that we are going to be using across our configuration in the project.

Reviewer: ```suggestion
Now let's add some variables to the `project.garden.yml` that we are going to be using across our configuration.
```

```YAML
variables:
  # Variables for Docker Registry
  registryHostname: docker.io
  registryNamespace: your-registry-namespace # Username ex. shankyweb in Dockerhub
  # Variables for Kubernetes
  DEFAULT_NAMESPACE: default # Frontend will be deployed in this namespace.
  # Cert-Manager variables
  CERT_MANAGER_INSTALL_CRDS: true
  GENERATE_PROD_CERTS: false
  GENERATE_STG_CERTS: false
  CF_DNS_PROVIDER: cloudflare # This demo only supports Cloudflare at the moment
  CF_DOMAIN: [ your-domain.com ] # Has to be an array, even if you only have one domain
  CF_UPDATE_STRATEGY: sync
  CF_EMAIL: your-email@myemail.com
  CF_PROXIED: false
```

> The variables that are required to be modified according to your environment are:

Reviewer: ```suggestion
The variables that require modification are:
```

- `registryHostname`: Add your registry, in this example we are going to be using Docker Hub.
- `registryNamespace`: Your registry namespace, in Dockerhub this will be your username.
- `CF_DOMAIN`: Add the DNS you will be using for this project in the form of an array.
- `CF_EMAIL`: Add the email address that should get notifications from Letsencrypt/Cloudflare.

There is also one secret that we require you to have `exported` at the OS level.

To make secret management easier we use environment variables to provide the Cloudflare API Token. Make sure to have this variable exported as it's used across the project.

````bash
export CF_API_TOKEN="your-cloudflare-api-key"
````

> The other variables *should* be set with the values we are providing at this stage as they will configure the project preventing errors to happen, most of them are extra configuration to add another level of customization to this example.

Reviewer: ```suggestion
Keep the other variables set to default: they will smooth out the deployment process and prevent errors.
```

name: cert-manager-ext-dns
defaultEnvironment: prod

environments:
  - name: prod
    production: true
    defaultNamespace: ${var.DEFAULT_NAMESPACE}
    variables:
      base-hostname: ${var.CF_DOMAIN[0]}

providers:
  - name: kubernetes
    environments: [prod]
    context: arn:aws:eks:eu-central-1:431328314483:cluster/cluster-8
    namespace: ${var.DEFAULT_NAMESPACE}
    setupIngressController: nginx
    buildMode: kaniko
    kaniko:
      namespace: null
    # tlsCertificates:
    #   - name: staging-cert
    #     secretRef:
    #       name: staging-cert
    deploymentRegistry:
      hostname: "${var.registryHostname}"
      namespace: "${var.registryNamespace}"
    imagePullSecrets:
    - name: regcred
      namespace: default

variables:
  # Variables for Docker Registry
  registryHostname: docker.io
  registryNamespace: shankyweb
  # Variables for Kubernetes
  DEFAULT_NAMESPACE: default
  # Cert-Manager variables
  CERT_MANAGER_INSTALL_CRDS: true
  GENERATE_PROD_CERTS: false
  GENERATE_STG_CERTS: false
  # ExternalDNS variables
  CF_DNS_PROVIDER: cloudflare
  CF_DOMAIN: [ shankyjs.com ]
  CF_UPDATE_STRATEGY: sync
  CF_EMAIL: hello@shankyjs.com
  CF_PROXIED: false

```

> **Note:** Variables are going to be start to make sense after we start building our Helm Modules for `cert-manager` and `external-dns`.

Reviewer: ```suggestion
**Note:** Variables are going to start to make sense after we build our Helm modules for `cert-manager` and `external-dns`.
```

### Creating our Helm Modules 🔨

Now we create [Helm modules](https://docs.garden.io/reference/module-types/helm) responsible for deploying our Helm Charts to our Kubernetes Cluster.

#### Configuring Cert-Manager module

Let's edit our `./garden.yml` file to define a Helm module that handles the installation of the `cert-manager` operator, in charge of creating and renewing our certificates.

> For this Helm Chart we only require setting 1 value, `installCRDs`, in charge of installing the [Custom Resource Definitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) in our Kubernetes Cluster.

Reviewer: ```suggestion
For this Helm chart we only require setting 1 value, `installCRDs`, in charge of installing the [Custom Resource Definitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) in our Kubernetes Cluster.
```

Also, something to note is that we are starting to use the environment variables that we set previously at the `garden.project.yml` level.

````yaml
---
# Cert-Manager Helm Chart configuration
kind: Module

> type: helm

Reviewer 2: this tutorial contains a lot of text, and a lot of yaml code. I think it will be hard to maintain and test, and the document might degrade over time and stop being helpful as it becomes outdated, and at the same time it's hard to maintain because it can't be tested by running a few commands.

I think there is some duplication between the example and this tutorial. How about walking the user through using the example, instead of reproducing the example completely in the form of a tutorial? 

This would get users up to speed more quickly, and make this guide easier to maintain.

Reviewer: I suggested Shanky in-line code for readability. I agree the document is lengthy because we set up both ExternalDNS and cert-manager but I don't want to risk users losing context. An alternative could be to split the guides so that ExternalDNS is set up in another section from cert-manager and they reference each other.

Reviewer: I'm also confused by the two documents under `docs/k8s-plugins/remote-k8s/ingress-and-dns.md` and `docs/advanced/cert-manager-integration.md`: a quick glance tells me there is a lot of duplication as @stefreak suggests. I suggest just a quick mention on the ingress and TLS page that points to where the cert-manager guide is located so users can "Choose Their Own Adventure" whether they proceed with the documentation on cloud cert provisioning or manage it themselves using the cert-manager guide.

Reviewer 2: I would try to avoid inline code that might get outdated, e.g. due to new helm chart versions that break backwards compatibility. 
Can we also explain the high level concepts and the steps required without too much inline code?

I'm afraid that the tutorial and the example diverge as the code evolve, leaving users who follow the tutorial a potential bad user experience because we weren't able to continuously test the snippets carefully.

For example, what will happen pretty soon is we'll need to port this example to 0.13.

If some snippets are required, that's fine– but I feel right now almost all the code is duplicated across the tutorial and the example.

name: cert-manager
namespace: cert-manager
description: This module installs cert-manager in our Kubernetes Cluster.
repo: https://charts.jetstack.io
chart: "cert-manager"
version: v1.11.0
chartPath: .
values:
  installCRDs: ${var.CERT_MANAGER_INSTALL_CRDS}
---
````

#### Configuring ExternalDNS module ☁️

ExternalDNS is the operator that is going to be in charge of creating our DNS entries into our CloudFlare account in an automated way.

> The combination of ExternalDNS + Cert-Manager is a powerful and production ready solution to manage and deploy certificates. 🚀

Reviewer: ```suggestion
The combination of ExternalDNS + cert-manager is a powerful and production ready solution to manage and deploy certificates. 🚀
```

Add the following to your `garden.yml`:

````yaml
# ExternalDNS Helm Chart
kind: Module
type: helm
name: external-dns
namespace: external-dns
description: This module installs external-dns in our Kubernetes Cluster.
repo: https://charts.bitnami.com/bitnami
chart: "external-dns"
version: 6.13.3
chartPath: .
values:
  provider: ${var.CF_DNS_PROVIDER}
  domainFilters: ${var.CF_DOMAIN}
  policy: ${var.CF_UPDATE_STRATEGY}
  cloudflare:
    email: ${var.CF_EMAIL}
    proxied: ${var.CF_PROXIED}
    apiToken: ${local.env.CF_API_TOKEN}
````

This module requires more configuration because it needs the correct Zone Details and values. For a full list of the possible values you can set see the [Helm Chart documentation](https://artifacthub.io/packages/helm/bitnami/external-dns?modal=values)

#### Configuring cluster-issuers module

> This Helm module is going to be different than the others we previously configured, because this Helm chart is intended to deploy the Custom Resources that we are going to need in order to interact with this demo, and the Helm Chart is going to be living in the same repository we are executing this demo.

Reviewer: ```suggestion
This Helm module is going to be different than the others we previously configured, because this Helm chart is intended to deploy the Custom Resources that we are going to need in order to interact with this demo, and the Helm chart is going to be living in the same repository we are executing this demo from.
```

> At this point we are just going to add the configuration of the Helm Module for Garden, we will create the chart in the next section of this tutorial.

Reviewer: ```suggestion
At this point we are just going to add the configuration of the Helm module for Garden: we will create the chart in the next section of this tutorial.
```

repo: https://charts.bitnami.com/bitnami
chart: "external-dns"
version: 6.13.3
chartPath: .
values: # To check a full list of values, see https://artifacthub.io/packages/helm/bitnami/external-dns
  provider: ${var.CF_DNS_PROVIDER}
  domainFilters: ${var.CF_DOMAIN}
  policy: ${var.CF_UPDATE_STRATEGY}
  cloudflare:
    email: ${var.CF_EMAIL}
    proxied: ${var.CF_PROXIED}
    apiToken: ${local.env.CF_API_TOKEN}
---
# Cluster-issuers module configuration
kind: Module
type: helm
name: cluster-issuers
namespace: default
description: This module installs cluster-issuers in our Kubernetes Cluster.
chartPath: charts/cluster-issuers
dependencies:
  - cert-manager # This module depends on the cert-manager module (we need the CRDs to be installed)
values:
  cloudflare:
    email: ${var.CF_EMAIL}
    apiToken: ${local.env.CF_API_TOKEN}
    cfDomain: ${var.CF_DOMAIN[0]} # For this example, at the moment we only support one domain.
  generateStgCert: ${var.GENERATE_STG_CERTS}
  generateProdCert: ${var.GENERATE_PROD_CERTS}

````

Make sure to review if your file matches this and correct if it's needed.

### Creating the cluster-issuers Helm Chart

> Now that we have ExternalDNS and Cert-manager charts configured, we still have to configure our Cert-manager Custom Resources.

Reviewer: ```suggestion
Now that we have ExternalDNS and cert-manager charts configured, we still have to configure our cert-manager Custom Resources.
```

> In order to generate our certificates we need some components from cert-manager applied into our Kubernetes Cluster, these components are the following:

Reviewer: ```suggestion
In order to generate our certificates we need some components from cert-manager applied into our Kubernetes cluster:
```

- ClusterIssuers: A Cluster issuer is in charge of identifying which Certificate Authority (CA) cert-manager will use to issue a certificate.
- Certificates: Represents a human readable definition of a certificate request that is created by an issuer. In order to generate a certificate you need a ClusterIssuer or an Issuer first.
- Cloudflare-API-Token Secret: This Kubernetes secret is going to be used so `cert-manager` can do the DNS01 challenge using the CloudFlare API Token.

> To install those components you can either follow [this guide](https://cert-manager.io/docs/configuration/acme/dns01/cloudflare/) from cert-manager to create the ClusterIssuer and needed secrets.

Reviewer: ```suggestion
To install those components you can follow [this guide](https://cert-manager.io/docs/configuration/acme/dns01/cloudflare/) from cert-manager to create the ClusterIssuer and needed secrets.
```

When you use "either" in a sentence you must offer the alternative inside the same sentence.

> If you want to have this fully automated with Garden you'll need to create a Helm Chart (or any sort of manifest applying mechanism like Kustomize) to create and apply those manifests to your cluster.

Reviewer: ```suggestion
To fully automate certificate issuance and manage cert-manager with Garden, you'll need to create a Helm chart to create and apply those manifests to your cluster.
```

> If you decide to go with the automated way we created a custom Helm Chart that is stored in our [cert-manager-ext-dns](https://github.com/garden-io/garden/tree/main/examples/cert-manager-ext-dns/charts/cluster-issuers) example folder.

Reviewer: ```suggestion
If you decide to go with the automated way we created a custom Helm chart that is stored in our [cert-manager-ext-dns](https://github.com/garden-io/garden/tree/main/examples/cert-manager-ext-dns/charts/cluster-issuers) example folder.
```

> To use it just proceed to clone the repository and move the charts folder into your own repository

Reviewer: ```suggestion
To use it, clone the repository and move the charts folder into your own repository.
```

````bash
git clone git@github.com:garden-io/garden.git
cp -f examples/cert-manager-ext-dns/charts ./charts
````

After you imported the `charts` folder make sure that you have the configuration for the module in the `garden.yml` file.

````yaml
---
# Cluster-issuers module configuration
kind: Module
type: helm
name: cluster-issuers
namespace: default
description: This module installs cluster-issuers in our Kubernetes Cluster.
chartPath: charts/cluster-issuers
dependencies:
  - cert-manager # This module depends on the cert-manager module (we need the CRDs to be installed)
values:
  cloudflare:
    email: ${var.CF_EMAIL}
    apiToken: ${local.env.CF_API_TOKEN}
    cfDomain: ${var.CF_DOMAIN[0]} # For this example, at the moment we only support one domain.
  generateStgCert: ${var.GENERATE_STG_CERTS}
  generateProdCert: ${var.GENERATE_PROD_CERTS}

````

After either importing our Helm Chart or creating the ClusterIssuers/Certificates manually, you are ready to move to the next step.

### Creating our Frontend

> To showcase our certificates we decided to use a sample react-application, the code for that React Application is hosted in our examples folder in the main `Garden` repository [cert-manager-ext-dns](https://github.com/garden-io/garden/tree/main/examples/cert-manager-ext-dns/frontend).

Reviewer: ```suggestion
To showcase our certificates we use a simple React application. The code for that React application is hosted in our `examples` folder in the main Garden repository [cert-manager-ext-dns](https://github.com/garden-io/garden/tree/main/examples/cert-manager-ext-dns/frontend).
```

`examples` refers to a file path so that is surrounded by backticks. Garden, when capitalized refers to Garden the software tool, company and ecosystem and is not backticked. Sometimes we wish to refer to `garden` the binary or when instructing users to e.g. `garden deploy`: that's a good place to format `garden` as code.

You can simply clone that repository; and move the code to the folder called `frontend` in your example.

#### Cloning the existing repo

The following commands needs to be executed from your root (next to where the `project.garden.yml` file is.)

````bash
git clone git@github.com:garden-io/garden.git
cp -f examples/cert-manager-ext-dns/frontend ./frontend
````

This folder has the `garden.yml` and `Dockerfile` ready, so you are ready to go to the next step.

> The most important part of this configuration is the ingresses block, as it's where we set the hostname for the application `Ex. react.shankyjs.com`.

Reviewer: ```suggestion
The most important part of this configuration is the ingresses block, as it's where we set the hostname for the application `e.g. react.shankyjs.com`.
```

      ℹ Run with --force-refresh to force a refresh of provider statuses.
   ✔ graph                     → Resolving 11 modules... → Done
   ✔ kubernetes                → Configuring... → Ready
   ℹ Run with --force-refresh to force a refresh of provider statuses.
✔ graph                     → Resolving 4 modules... → Done
✔ cluster-issuers           → Building version v-7e466ae4e3... → Done (took 0 sec)
✔ cert-manager              → Building version v-0b43e69fd6... → Done (took 0.6 sec)
✔ external-dns              → Building version v-ec953ac304... → Done (took 1.1 sec)
✔ react-app                 → Getting build status for v-8323dd9f4c... → Already built
✔ cert-manager              → Deploying version v-010a92e8da... → Done (took 37 sec)
   ℹ cert-manager              → Resources ready
✔ external-dns              → Deploying version v-11221c2d8c... → Done (took 12.9 sec)
   ℹ external-dns              → Resources ready
✔ cluster-issuers           → Deploying version v-e67a273f4c... → Already deployed
✔ frontend                  → Deploying version v-ac87f68175... → Already deployed
   Ingress: http://react.shankyjs.com

Done! ✔️
````

> Our website is now up, but without a certificate; because ExternalDNS already provisioned our DNS records, but we still haven't generated any Certificate for our frontend application.

Reviewer: ```suggestion
ExternalDNS already provisioned our DNS records but we still haven't generated any certificate for our frontend application. Let's issue those now.
```

![Frontend application without HTTPs certificate](https://res.cloudinary.com/djp21wtxm/image/upload/v1676712587/i1600x744-DlhjPIr3f0XI_aut50k.png)

### Issuing our first certificate (Staging)

A good practice whenever experimenting with certificates is to use the Staging Certificate Authority to issue our first certificates. By doing this we can easily debug and identify any issues with our certificates without affecting our limited `Production` limits.

> In order to generate our first staging-certificate we need to do the following:

Reviewer: ```suggestion
In order to generate our first staging certificate, we need to do the following:
```

    tlsCertificates: # You can start testing this with the staging certificates, but you should use production certificates in production
      - name: staging-cert
        secretRef:
          name: staging-cert
    ...
````

Let's deploy one more time with `garden deploy --yes`

Now go back to your React Application using your browser: the URL is react.${your-domain}.

And! Surprise 🎉

![Your connection is not private alert](https://res.cloudinary.com/djp21wtxm/image/upload/v1676834933/i1425x1036-pyYqS8azWa2P_ccsaez.png)

If you see the "Your connection is not private" alert, this means that we are on track, because this alert means that our browser doesn't trust the certificate generated by the Staging CA, however it means that the Certificate was provisioned successfully in our Ingress.

If you click "Advanced" in your web browser and then proceed to `react.${your-domain-termination}`" you will be able to view your fancy Hello Garden 🌸 landing page using the staging Let's Encrypt certificate.

![Describing our certificate (Staging)](https://res.cloudinary.com/djp21wtxm/image/upload/v1676835016/i1600x1085-R-nlUNxT6bML_eangbo.png)

### Issuing our Production Certificate 🎉

> This is the last part of this demo: run this part after you've finished your experiments with staging. Remember that the API Limits and Quotas from Production Let's Encrypt are lower so you shouldn't be generating excessive amounts of certificates in a short-span of time.

Reviewer: ```suggestion
This is the last part of this demo: run this part after you've finished your experiments with staging. Remember that the API limits and quotas from production Let's Encrypt are lower so you shouldn't be generating excessive amounts of certificates in a short-span of time.
```

Why was Limits and Quota capitalized?
